### ISSUE TITLE
App controller methods are being called multiple times

### ISSUE THREAD

**Issue by Sharsie** [soberwp/controller#57]
These tests are performed on a homepage

I have the following structure:

app/controllers/App.php
app/controllers/Home.php

```
...
class App extends Controller {
    public function app_test()
    {
         error_log('called App::test');
    }
}
```

```
...
class Home extends Controller {
    public function home_test()
    {
         error_log('called Home::test');
    }
}
```

the result of the error log for a single request is:
```
called App::test
called App::test
called Home::test
called App::test
```

So far I have been debugging I found out the following:
`\Sober\Controller\loader` function adds filters for
`sage/template/app-data/data` (App controller)
`sage/template/home-data/data` (Home controller)

Sage's `template_include` filter then runs the above filters (it actually calls several other based on the body classes)

the filter then creates a new controller [for each of the filters it added](https://github.com/soberwp/controller/blob/93aa3ccd869b6092eefb941e342382ce7e3b00db/controller.php#L14-L16), in our case it can be translated as

```
    $controller = new App\Controllers\App(); // First instance of App being created
    $controller->__setup();
    return array_merge($loader->getAppData(), $loader->getPostData(), $controller->__setTreeData($data), $controller->__getData());
```
and

```
    $controller = new App\Controllers\Home(); // First instance of Home being created
    $controller->__setup();
    return array_merge($loader->getAppData(), $loader->getPostData(), $controller->__setTreeData($data), $controller->__getData());
```

So to clarify: We call the `__setup` method on the controller once for App and then once for Home.

The problem is in the array_merge when [$loader calls `getAppData`](https://github.com/soberwp/controller/blob/93aa3ccd869b6092eefb941e342382ce7e3b00db/src/Loader.php#L121-L129) which in turn calls `__setup` on App again

This results in the following calls:
`__setup` from the `sage/template/app-data/data` filter on App class (first App instance)
`__setup` from `getAppData` on `new $this->instances['app']()` from `sage/template/app-data/data` filter (second instance of App being created)
`__setup` from the `sage/template/home-data/data` filter on Home class (first instance of Home)
`__setup` from `getAppData` on `new $this->instances['app']()` from filter `sage/template/home-data/data` (third instance of App)

So I have a solution to migrate the controllers to singletons (not sure if this is viable as we would need protected contructors for a clean design and this might break already existing solutions based on the current controller pattern) - this could also make it harder for dependency injection if it was to be implemented later on

So if anyone has a better idea please share.

I will submit a PR for this in a bit in case nobody comes up with a better solution

**Comment by darrenjacoby** [comment]
@Sharsie 

I've done a lot of work here, using the Sage DI container to create the classes, removed the App data from the loop, etc. It was a fairly big rework and clean up on `controller.php` and `Loader.php`. As a result classes are only being created once now. 

This solves the above issue without extra code—it was part of the work to make dependency injection work for Controllers.

Thanks for your contribution, and would love for you to test the current solution once released to dev-master.

**Comment by Sharsie** [comment]
Perfect, thanks, I will have a look at it once it's out

**Comment by slackday** [comment]
Hi, did this make it into 2.0.0? Looks like it from the code but changelog isn't updated. I would also like to try 2.0.0 👍

**Comment by darrenjacoby** [comment]
@slackday yea this working as should now. I forgot to update the changelog, will do that.

**Comment by studio404pl** [comment]
@darrenjacoby is this fix will be released on 9.0.0 also (sage user here).
